In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")
import os

#### Loading all the excel file into jupyter notebook

In [2]:
# Checking details of Comapny's Order Report File 
orderrep=pd.read_excel("C:\data science\Cointab Final Assignment Data\Company X - Order Report.xlsx")

In [3]:
orderrep.head(10)

ExternOrderNo            SKU  Order Qty
0     2001827036  8904223818706        1.0
1     2001827036  8904223819093        1.0
2     2001827036  8904223819109        1.0
3     2001827036  8904223818430        1.0
4     2001827036  8904223819277        1.0
5     2001827036  GIFTBOX202002        1.0
6     2001827036  8904223818638        2.0
7     2001827036     SACHETS001        1.0
8     2001825261  8904223819024        4.0
9     2001825261  8904223819291        4.0

In [4]:
#Checking the size of file
orderrep.shape

(400, 3)

In [5]:
# Checking for duplicate entries
orderrep.duplicated().sum()

2

In [6]:
# Removing the duplicates
orderrep.drop_duplicates(inplace=True)

In [7]:
# Rechecking the shape
orderrep.shape

(398, 3)

Obs- We can see the duplicates have been removed

In [8]:
# Checking details of Customer Pincode File 
pincode=pd.read_excel("C:\data science\Cointab Final Assignment Data\Company X - Pincode Zones.xlsx")

In [9]:
pincode.head(10)

Warehouse Pincode  Customer Pincode Zone
0             121003            507101    d
1             121003            486886    d
2             121003            532484    d
3             121003            143001    b
4             121003            515591    d
5             121003            326502    d
6             121003            208019    b
7             121003            140301    b
8             121003            396001    d
9             121003            711106    d

In [10]:
#Checking the size of file
pincode.shape

(124, 3)

In [11]:
# Checking for duplicate entries
pincode.duplicated().sum()

16

In [12]:
# Removing the duplicates
pincode.drop_duplicates(inplace=True)

In [13]:
# Rechecking the shape
pincode.shape

(108, 3)

Obs- We can see the duplicates have been removed

In [14]:
# Checking details of Comapny's SKU Master File 
skumaster=pd.read_excel("C:\data science\Cointab Final Assignment Data\Company X - SKU Master.xlsx")

In [15]:
skumaster.head(10)

SKU  Weight (g)
0  8904223815682         210
1  8904223815859         165
2  8904223815866         113
3  8904223815873          65
4  8904223816214         120
5  8904223816665         102
6  8904223817273          65
7  8904223817334         170
8  8904223817501         350
9  8904223818430         165

In [16]:
#Checking the size of file
skumaster.shape

(66, 2)

In [17]:
# Checking for duplicate entries
skumaster.duplicated().sum()

1

In [18]:
# Removing the duplicates
skumaster.drop_duplicates(inplace=True)

In [19]:
# Rechecking the shape
skumaster.shape

(65, 2)

Obs- We can see the duplicate entries have been removed

In [20]:
#Checking details of Comapny's Order Report File 
invoice=pd.read_excel("C:\data science\Cointab Final Assignment Data\Courier Company - Invoice.xlsx")

In [21]:
invoice.head(10)

AWB Code    Order ID  Charged Weight  Warehouse Pincode  \
0  1091117222124  2001806232            1.30             121003   
1  1091117222194  2001806273            1.00             121003   
2  1091117222931  2001806408            2.50             121003   
3  1091117223244  2001806458            1.00             121003   
4  1091117229345  2001807012            0.15             121003   
5  1091117229555  2001806686            0.15             121003   
6  1091117229776  2001806885            1.00             121003   
7  1091117323112  2001807058            1.15             121003   
8  1091117323812  2001807186            0.50             121003   
9  1091117324206  2001807290            0.50             121003   

   Customer Pincode Zone Type of Shipment  Billing Amount (Rs.)  
0            507101    d  Forward charges                 135.0  
1            486886    d  Forward charges                  90.2  
2            532484    d  Forward charges                 224.6  
3            143001    b  Forward charges                  61.3  
4            515591    d  Forward charges                  45.4  
5            326502    d  Forward charges                  45.4  
6            208019    b  Forward charges                  61.3  
7            140301    b  Forward charges                  89.6  
8            396001    d  Forward charges                  45.4  
9            711106    d  Forward charges                  45.4

In [22]:
#Checking the size of file
invoice.shape

(124, 8)

In [23]:
# Checking for duplicate entries
invoice.duplicated().sum()

0

Obs- No duplicate entries present in this file

In [24]:
# Checking details of Comapny's Order Report File (Have modified this file by bringing all the rates in one row)
rates=pd.read_excel("C:\data science\Cointab Final Assignment Data\Courier Company - Rates Mod.xlsx")

In [25]:
rates.head()

fwd_a_fixed  fwd_a_additional  fwd_b_fixed  fwd_b_additional  fwd_c_fixed  \
0         29.5              23.6           33              28.3         40.1   

   fwd_c_additional  fwd_d_fixed  fwd_d_additional  fwd_e_fixed  \
0              38.9         45.4              44.8         56.6   

   fwd_e_additional  rto_a_fixed  rto_a_additional  rto_b_fixed  \
0              55.5         13.6              23.6         20.5   

   rto_b_additional  rto_c_fixed  rto_c_additional  rto_d_fixed  \
0              28.3         31.9              38.9         41.3   

   rto_d_additional  rto_e_fixed  rto_e_additional  
0              44.8         50.7              55.5

## Creating 1st Output file: Order Level Calculation 

The required fields/columns are to be included in the CSV/Excel file:

 ● OrderID

 ● AWBNumber
    
 ● Total weight as per X (KG)

 ● Weight slab as per X (KG)
    
 ● Total weight as per Courier Company (KG)

 ● Weight slab charged by Courier Company (KG)
    
 ● Delivery Zone as per X

 ● Delivery Zone charged by Courier Company
    
 ● Expected Charge as per X (Rs.)

 ● Charges Billed by Courier Company (Rs.)
    
 ● Difference Between Expected Charges and Billed Charges (Rs.)

#### *  Creating a new data frame "data1" and adding the fields/column one by one

In [26]:
data1=pd.DataFrame()
data1[["OrderID","AWBNumber"]]=invoice[['Order ID',"AWB Code"]].copy()

In [27]:
data1.head()

OrderID      AWBNumber
0  2001806232  1091117222124
1  2001806273  1091117222194
2  2001806408  1091117222931
3  2001806458  1091117223244
4  2001807012  1091117229345

In [28]:
data1.shape

(124, 2)

Obs- We can see 2 columns have been added

#### * To get the next column "Total weight as per X (KG)", we need to follow the following steps:
* 1. Merging the two tables Order Report with SKU Master order
* 2. This will lead to mapping order ID's , SKU's and weight in grams
* 3. Finding weight as per the no. of orderss and thus gaining final weight
* 4. Also since the Order Report have multiple entries with same SKU's, the merging will take all those entries.
* 5. Further bringing them as per Order ID, by using Group By function

In [29]:
merge_SKU=orderrep.merge(skumaster,on='SKU',how='left')

In [30]:
merge_SKU

ExternOrderNo            SKU  Order Qty  Weight (g)
0       2001827036  8904223818706        1.0         127
1       2001827036  8904223819093        1.0         150
2       2001827036  8904223819109        1.0         100
3       2001827036  8904223818430        1.0         165
4       2001827036  8904223819277        1.0         350
..             ...            ...        ...         ...
393     2001806229  8904223818942        1.0         133
394     2001806229  8904223818850        1.0         240
395     2001806226  8904223818850        2.0         240
396     2001806210  8904223816214        1.0         120
397     2001806210  8904223818874        1.0         100

[398 rows x 4 columns]

In [31]:
merge_SKU['Tot Wt']=merge_SKU['Order Qty']*merge_SKU['Weight (g)'].copy()

In [32]:
merge_SKU.head(10)

ExternOrderNo            SKU  Order Qty  Weight (g)  Tot Wt
0     2001827036  8904223818706        1.0         127   127.0
1     2001827036  8904223819093        1.0         150   150.0
2     2001827036  8904223819109        1.0         100   100.0
3     2001827036  8904223818430        1.0         165   165.0
4     2001827036  8904223819277        1.0         350   350.0
5     2001827036  GIFTBOX202002        1.0         500   500.0
6     2001827036  8904223818638        2.0         137   274.0
7     2001827036     SACHETS001        1.0          10    10.0
8     2001825261  8904223819024        4.0         112   448.0
9     2001825261  8904223819291        4.0         112   448.0

Obs- We can see where there are 2 or more orders, the tot wt field has been modified 

In [33]:
# Checking the shape
merge_SKU.shape

(398, 5)

Obs- We can see redundant data has been created, and to remove teh redundancy we will be using group by below

In [34]:
# Grouping the SKU's wrt to Order No
merge_SKU.groupby('ExternOrderNo')['Tot Wt'].sum().to_frame()

Tot Wt
ExternOrderNo        
2001806210      220.0
2001806226      480.0
2001806229      500.0
2001806232     1302.0
2001806233      245.0
...               ...
2001821995      477.0
2001822466     1376.0
2001823564      672.0
2001825261     1557.0
2001827036     1676.0

[124 rows x 1 columns]

In [35]:
data1=data1.merge(merge_SKU.groupby('ExternOrderNo')['Tot Wt'].sum().to_frame(),left_on='OrderID',right_on='ExternOrderNo')

In [36]:
data1

OrderID      AWBNumber  Tot Wt
0    2001806232  1091117222124  1302.0
1    2001806273  1091117222194   615.0
2    2001806408  1091117222931  2265.0
3    2001806458  1091117223244   700.0
4    2001807012  1091117229345   240.0
..          ...            ...     ...
119  2001812941  1091118551656   500.0
120  2001809383  1091117614452   607.0
121  2001820978  1091120922803   515.0
122  2001811475  1091121844806   689.0
123  2001811305  1091121846136   750.0

[124 rows x 3 columns]

In [37]:
data1.shape

(124, 3)

Obs- We can see the data has come in its original shape and the weight has also been added

In [38]:
# Since the asked entry of Total weight as per X (KG) is asked in KGS , coverting the weight in grams to KG

In [39]:
data1['Total weight as per X (KG)']=data1['Tot Wt']/1000
data1.shape

(124, 4)

In [40]:
data1.drop('Tot Wt',axis=1,inplace=True)
data1.head()

OrderID      AWBNumber  Total weight as per X (KG)
0  2001806232  1091117222124                       1.302
1  2001806273  1091117222194                       0.615
2  2001806408  1091117222931                       2.265
3  2001806458  1091117223244                       0.700
4  2001807012  1091117229345                       0.240

Obs- We have got our third field/column

#### * Now to get the 4th column i.e "Weight slab as per X (KG)", we have observed it follows a increemental step size of 0.5, and thus applying the math formula ceiling to first convert into its nearest integer first and then dividing by 2

In [41]:
data1['Weight slab as per X (KG)']=np.ceil(data1['Total weight as per X (KG)']*2)/2 

In [42]:
data1.head(10)

OrderID      AWBNumber  Total weight as per X (KG)  \
0  2001806232  1091117222124                       1.302   
1  2001806273  1091117222194                       0.615   
2  2001806408  1091117222931                       2.265   
3  2001806458  1091117223244                       0.700   
4  2001807012  1091117229345                       0.240   
5  2001806686  1091117229555                       0.240   
6  2001806885  1091117229776                       0.420   
7  2001807058  1091117323112                       1.168   
8  2001807186  1091117323812                       0.500   
9  2001807290  1091117324206                       0.500   

   Weight slab as per X (KG)  
0                        1.5  
1                        1.0  
2                        2.5  
3                        1.0  
4                        0.5  
5                        0.5  
6                        0.5  
7                        1.5  
8                        0.5  
9                        0.5

Obs- We have obtained the weight slab as per the X company 

#### * Now, to add the 5th column: "Total weight as per Courier Company (KG)" , we have to add "Charged Weight" field from invoice table

In [43]:
invoice.rename({'Order ID':'OrderID'},axis=1,inplace=True)

In [44]:
invoice.rename({'AWB Code':'AWBNumber'},axis=1, inplace=True)

In [45]:
data1=data1.merge(invoice[['Charged Weight','OrderID']],on='OrderID')

In [46]:
data1

OrderID      AWBNumber  Total weight as per X (KG)  \
0    2001806232  1091117222124                       1.302   
1    2001806273  1091117222194                       0.615   
2    2001806408  1091117222931                       2.265   
3    2001806458  1091117223244                       0.700   
4    2001807012  1091117229345                       0.240   
..          ...            ...                         ...   
119  2001812941  1091118551656                       0.500   
120  2001809383  1091117614452                       0.607   
121  2001820978  1091120922803                       0.515   
122  2001811475  1091121844806                       0.689   
123  2001811305  1091121846136                       0.750   

     Weight slab as per X (KG)  Charged Weight  
0                          1.5            1.30  
1                          1.0            1.00  
2                          2.5            2.50  
3                          1.0            1.00  
4                          0.5            0.15  
..                         ...             ...  
119                        0.5            0.73  
120                        1.0            0.50  
121                        1.0            0.50  
122                        1.0            0.50  
123                        1.0            0.50  

[124 rows x 5 columns]

Obs- The new column is added, but it is asked with the following name-"Total weight as per Courier Company (KG)" and therefore renamig it

In [47]:
data1.rename({'Charged Weight':'Total weight as per Courier Company (KG)'},axis=1,inplace=True)

In [48]:
data1.head(10)

OrderID      AWBNumber  Total weight as per X (KG)  \
0  2001806232  1091117222124                       1.302   
1  2001806273  1091117222194                       0.615   
2  2001806408  1091117222931                       2.265   
3  2001806458  1091117223244                       0.700   
4  2001807012  1091117229345                       0.240   
5  2001806686  1091117229555                       0.240   
6  2001806885  1091117229776                       0.420   
7  2001807058  1091117323112                       1.168   
8  2001807186  1091117323812                       0.500   
9  2001807290  1091117324206                       0.500   

   Weight slab as per X (KG)  Total weight as per Courier Company (KG)  
0                        1.5                                      1.30  
1                        1.0                                      1.00  
2                        2.5                                      2.50  
3                        1.0                                      1.00  
4                        0.5                                      0.15  
5                        0.5                                      0.15  
6                        0.5                                      1.00  
7                        1.5                                      1.15  
8                        0.5                                      0.50  
9                        0.5                                      0.50

Obs- we can now say the 5th column has been added

#### * Adding the 6th column-"Weight slab charged by Courier Company (KG)"

In [49]:
data1['Weight slab charged by Courier Company (KG)']=np.ceil(data1['Total weight as per Courier Company (KG)']*2)/2

In [50]:
data1.head()

OrderID      AWBNumber  Total weight as per X (KG)  \
0  2001806232  1091117222124                       1.302   
1  2001806273  1091117222194                       0.615   
2  2001806408  1091117222931                       2.265   
3  2001806458  1091117223244                       0.700   
4  2001807012  1091117229345                       0.240   

   Weight slab as per X (KG)  Total weight as per Courier Company (KG)  \
0                        1.5                                      1.30   
1                        1.0                                      1.00   
2                        2.5                                      2.50   
3                        1.0                                      1.00   
4                        0.5                                      0.15   

   Weight slab charged by Courier Company (KG)  
0                                          1.5  
1                                          1.0  
2                                          2.5  
3                                          1.0  
4                                          0.5

#### * For adding the 7th column-"Delivery Zone as per X" , we should ideally take zone column from "Invoice Courier" file and merge with data1, however we can see there is dicrepancy in it, as the zones are not alligning with pincodes provided in Pincode file.
#### * Therefore creating a temporary variable to do the mapping

In [51]:
Temp=invoice.merge(pincode,on=['Customer Pincode','Warehouse Pincode'],how='right')
Temp

AWBNumber     OrderID  Charged Weight  Warehouse Pincode  \
0    1091117222124  2001806232            1.30             121003   
1    1091117222194  2001806273            1.00             121003   
2    1091117222931  2001806408            2.50             121003   
3    1091117223244  2001806458            1.00             121003   
4    1091117229345  2001807012            0.15             121003   
..             ...         ...             ...                ...   
119  1091117904860  2001811039            0.68             121003   
120  1091121846136  2001811305            0.50             121003   
121  1091118551656  2001812941            0.73             121003   
122  1091117614452  2001809383            0.50             121003   
123  1091120922803  2001820978            0.50             121003   

     Customer Pincode Zone_x         Type of Shipment  Billing Amount (Rs.)  \
0              507101      d          Forward charges                 135.0   
1              486886      d          Forward charges                  90.2   
2              532484      d          Forward charges                 224.6   
3              143001      b          Forward charges                  61.3   
4              515591      d          Forward charges                  45.4   
..                ...    ...                      ...                   ...   
119            302020      d          Forward charges                  90.2   
120            302020      d          Forward charges                  45.4   
121            325207      d          Forward charges                  90.2   
122            303702      d  Forward and RTO charges                  86.7   
123            313301      d          Forward charges                  45.4   

    Zone_y  
0        d  
1        d  
2        d  
3        b  
4        d  
..     ...  
119      b  
120      b  
121      b  
122      b  
123      b  

[124 rows x 9 columns]

Obs- Here we can see the zone field from Pincode file is added at right

In [52]:
# Making the merge on Data1, and also renaming AWB Code to AWB number as data1 has that particular name
Temp.rename({'AWB Code':'AWBNumber'},axis=1,inplace=True)

In [53]:
data1=data1.merge(Temp[['Zone_y','AWBNumber']],on='AWBNumber')

In [54]:
# However , we want it with the name-"Delivery Zone as per X" and thus renaming it
data1.rename({'Zone_y':'Delivery Zone as per X'},axis=1,inplace=True)

In [55]:
data1

OrderID      AWBNumber  Total weight as per X (KG)  \
0    2001806232  1091117222124                       1.302   
1    2001806273  1091117222194                       0.615   
2    2001806408  1091117222931                       2.265   
3    2001806458  1091117223244                       0.700   
4    2001807012  1091117229345                       0.240   
..          ...            ...                         ...   
119  2001812941  1091118551656                       0.500   
120  2001809383  1091117614452                       0.607   
121  2001820978  1091120922803                       0.515   
122  2001811475  1091121844806                       0.689   
123  2001811305  1091121846136                       0.750   

     Weight slab as per X (KG)  Total weight as per Courier Company (KG)  \
0                          1.5                                      1.30   
1                          1.0                                      1.00   
2                          2.5                                      2.50   
3                          1.0                                      1.00   
4                          0.5                                      0.15   
..                         ...                                       ...   
119                        0.5                                      0.73   
120                        1.0                                      0.50   
121                        1.0                                      0.50   
122                        1.0                                      0.50   
123                        1.0                                      0.50   

     Weight slab charged by Courier Company (KG) Delivery Zone as per X  
0                                            1.5                      d  
1                                            1.0                      d  
2                                            2.5                      d  
3                                            1.0                      b  
4                                            0.5                      d  
..                                           ...                    ...  
119                                          1.0                      b  
120                                          0.5                      b  
121                                          0.5                      b  
122                                          0.5                      e  
123                                          0.5                      b  

[124 rows x 7 columns]

Obs- We have got our 7th column

#### * Now to add the 8th column i.e "Delivery Zone charged by Courier Company", we again have to take data from the Temp variable created earlier

In [56]:
data1=data1.merge(Temp[['Zone_x','AWBNumber']],on='AWBNumber')

In [57]:
# Renaming as per the requirement
data1.rename({'Zone_x':'Delivery Zone charged by Courier Company'},axis=1,inplace=True)

In [58]:
data1

OrderID      AWBNumber  Total weight as per X (KG)  \
0    2001806232  1091117222124                       1.302   
1    2001806273  1091117222194                       0.615   
2    2001806408  1091117222931                       2.265   
3    2001806458  1091117223244                       0.700   
4    2001807012  1091117229345                       0.240   
..          ...            ...                         ...   
119  2001812941  1091118551656                       0.500   
120  2001809383  1091117614452                       0.607   
121  2001820978  1091120922803                       0.515   
122  2001811475  1091121844806                       0.689   
123  2001811305  1091121846136                       0.750   

     Weight slab as per X (KG)  Total weight as per Courier Company (KG)  \
0                          1.5                                      1.30   
1                          1.0                                      1.00   
2                          2.5                                      2.50   
3                          1.0                                      1.00   
4                          0.5                                      0.15   
..                         ...                                       ...   
119                        0.5                                      0.73   
120                        1.0                                      0.50   
121                        1.0                                      0.50   
122                        1.0                                      0.50   
123                        1.0                                      0.50   

     Weight slab charged by Courier Company (KG) Delivery Zone as per X  \
0                                            1.5                      d   
1                                            1.0                      d   
2                                            2.5                      d   
3                                            1.0                      b   
4                                            0.5                      d   
..                                           ...                    ...   
119                                          1.0                      b   
120                                          0.5                      b   
121                                          0.5                      b   
122                                          0.5                      e   
123                                          0.5                      b   

    Delivery Zone charged by Courier Company  
0                                          d  
1                                          d  
2                                          d  
3                                          b  
4                                          d  
..                                       ...  
119                                        d  
120                                        d  
121                                        d  
122                                        b  
123                                        d  

[124 rows x 8 columns]

Obs- We can see 8th column has been added

#### * Now for adding the 9th column " Expected Charge as per X (Rs.)", we have to take help of Rate chart provided and apply the formulas and calculate the final charges.

In [59]:
# Finding total count of Forward Charges and Forward with RTO charges

In [60]:
invoice['Type of Shipment'].value_counts()

Forward charges            109
Forward and RTO charges     15
Name: Type of Shipment, dtype: int64

Obs- We can see 15 orders have to go through RTO Charges

In [61]:
# Getting the types of shipment for all the zones
for i in rates.columns:
    x='{}={}'.format(i,rates[i][0])
    exec(x)
    print(i)

fwd_a_fixed
fwd_a_additional
fwd_b_fixed
fwd_b_additional
fwd_c_fixed
fwd_c_additional
fwd_d_fixed
fwd_d_additional
fwd_e_fixed
fwd_e_additional
rto_a_fixed
rto_a_additional
rto_b_fixed
rto_b_additional
rto_c_fixed
rto_c_additional
rto_d_fixed
rto_d_additional
rto_e_fixed
rto_e_additional


In [62]:
# Reusing the Temp varaible to find expected charges as per x
Temp=Temp.merge(data1[['Weight slab as per X (KG)','AWBNumber']],on='AWBNumber')

###### * To calculate the expected charges by X,when it is just 'Forward Charges' we will be using the below formula throughout the function:
 ###### charges= fixed+((np.ceil(Wt Slab/0.5)-1)*additional charges)
###### * To calculate the expected charges by X,when it is'Forward Charges +RTO' we will be using the below formula throughout the function:
 ###### charges=fixed+((np.ceil(Wt Slab/0.5)-1)*fwd_a_additional)+rto_fixed+((np.ceil(Wt Slab/0.5)-1)*rto_additional charges)''' 

In [63]:
def Calc(x):
    x=x.tolist()
    #print(x)
    if x[2]=='Forward charges':     #x[2], since we are passing 3 arguments to Calc function , and 3rd elements index will be 2
        if x[0]=='a':
            return fwd_a_fixed+((np.ceil(x[1]/0.5)-1)*fwd_a_additional)
        elif x[0]=='b':
            return fwd_b_fixed+((np.ceil(x[1]/0.5)-1)*fwd_b_additional)
        elif x[0]=='c':
            return fwd_c_fixed+((np.ceil(x[1]/0.5)-1)*fwd_c_additional)
        elif x[0]=='d':
            return fwd_d_fixed+((np.ceil(x[1]/0.5)-1)*fwd_d_additional)
        else:
            return fwd_e_fixed+((np.ceil(x[1]/0.5)-1)*fwd_e_additional)
    else:
        if x[0]=='a':
            return fwd_a_fixed+((np.ceil(x[1]/0.5)-1)*fwd_a_additional)+rto_a_fixed+((np.ceil(x[1]/0.5)-1)*rto_a_additional)
        elif x[0]=='b':
            return fwd_b_fixed+((np.ceil(x[1]/0.5)-1)*fwd_a_additional)+rto_b_fixed+((np.ceil(x[1]/0.5)-1)*rto_a_additional)
        elif x[0]=='c':
            return fwd_c_fixed+((np.ceil(x[1]/0.5)-1)*fwd_a_additional)+rto_c_fixed+((np.ceil(x[1]/0.5)-1)*rto_a_additional)
        elif x[0]=='d':
            return fwd_d_fixed+((np.ceil(x[1]/0.5)-1)*fwd_a_additional)+rto_d_fixed+((np.ceil(x[1]/0.5)-1)*rto_a_additional)
        else:
            return fwd_e_fixed+((np.ceil(x[1]/0.5)-1)*fwd_a_additional)+rto_e_fixed+((np.ceil(x[1]/0.5)-1)*rto_a_additional)

In [64]:
Temp

AWBNumber     OrderID  Charged Weight  Warehouse Pincode  \
0    1091117222124  2001806232            1.30             121003   
1    1091117222194  2001806273            1.00             121003   
2    1091117222931  2001806408            2.50             121003   
3    1091117223244  2001806458            1.00             121003   
4    1091117229345  2001807012            0.15             121003   
..             ...         ...             ...                ...   
119  1091117904860  2001811039            0.68             121003   
120  1091121846136  2001811305            0.50             121003   
121  1091118551656  2001812941            0.73             121003   
122  1091117614452  2001809383            0.50             121003   
123  1091120922803  2001820978            0.50             121003   

     Customer Pincode Zone_x         Type of Shipment  Billing Amount (Rs.)  \
0              507101      d          Forward charges                 135.0   
1              486886      d          Forward charges                  90.2   
2              532484      d          Forward charges                 224.6   
3              143001      b          Forward charges                  61.3   
4              515591      d          Forward charges                  45.4   
..                ...    ...                      ...                   ...   
119            302020      d          Forward charges                  90.2   
120            302020      d          Forward charges                  45.4   
121            325207      d          Forward charges                  90.2   
122            303702      d  Forward and RTO charges                  86.7   
123            313301      d          Forward charges                  45.4   

    Zone_y  Weight slab as per X (KG)  
0        d                        1.5  
1        d                        1.0  
2        d                        2.5  
3        b                        1.0  
4        d                        0.5  
..     ...                        ...  
119      b                        0.5  
120      b                        1.0  
121      b                        0.5  
122      b                        1.0  
123      b                        1.0  

[124 rows x 10 columns]

In [65]:
Temp['Expected Charge as per X (Rs.)']=Temp[['Zone_y','Weight slab as per X (KG)','Type of Shipment']].apply(Calc,axis=1)

In [66]:
# Merging the Expected Charge from Temp to data 1
data1=data1.merge(Temp[['Expected Charge as per X (Rs.)','AWBNumber']],on='AWBNumber')

In [67]:
data1

OrderID      AWBNumber  Total weight as per X (KG)  \
0    2001806232  1091117222124                       1.302   
1    2001806273  1091117222194                       0.615   
2    2001806408  1091117222931                       2.265   
3    2001806458  1091117223244                       0.700   
4    2001807012  1091117229345                       0.240   
..          ...            ...                         ...   
119  2001812941  1091118551656                       0.500   
120  2001809383  1091117614452                       0.607   
121  2001820978  1091120922803                       0.515   
122  2001811475  1091121844806                       0.689   
123  2001811305  1091121846136                       0.750   

     Weight slab as per X (KG)  Total weight as per Courier Company (KG)  \
0                          1.5                                      1.30   
1                          1.0                                      1.00   
2                          2.5                                      2.50   
3                          1.0                                      1.00   
4                          0.5                                      0.15   
..                         ...                                       ...   
119                        0.5                                      0.73   
120                        1.0                                      0.50   
121                        1.0                                      0.50   
122                        1.0                                      0.50   
123                        1.0                                      0.50   

     Weight slab charged by Courier Company (KG) Delivery Zone as per X  \
0                                            1.5                      d   
1                                            1.0                      d   
2                                            2.5                      d   
3                                            1.0                      b   
4                                            0.5                      d   
..                                           ...                    ...   
119                                          1.0                      b   
120                                          0.5                      b   
121                                          0.5                      b   
122                                          0.5                      e   
123                                          0.5                      b   

    Delivery Zone charged by Courier Company  Expected Charge as per X (Rs.)  
0                                          d                           135.0  
1                                          d                            90.2  
2                                          d                           224.6  
3                                          b                            61.3  
4                                          d                            45.4  
..                                       ...                             ...  
119                                        d                            33.0  
120                                        d                           100.7  
121                                        d                            61.3  
122                                        b                           112.1  
123                                        d                            61.3  

[124 rows x 9 columns]

Obs- We can see the 9th column is added

#### * To add the 10th column i.e "Charges Billed by Courier Company (Rs.)", we can use the "Billing Amount" field from invoice table

In [68]:
data1=data1.merge(invoice[['Billing Amount (Rs.)','AWBNumber']],on='AWBNumber')

In [69]:
data1.rename({'Billing Amount (Rs.)':'Charges Billed by Courier Company (Rs.)'},inplace=True,axis=1)

In [70]:
data1

OrderID      AWBNumber  Total weight as per X (KG)  \
0    2001806232  1091117222124                       1.302   
1    2001806273  1091117222194                       0.615   
2    2001806408  1091117222931                       2.265   
3    2001806458  1091117223244                       0.700   
4    2001807012  1091117229345                       0.240   
..          ...            ...                         ...   
119  2001812941  1091118551656                       0.500   
120  2001809383  1091117614452                       0.607   
121  2001820978  1091120922803                       0.515   
122  2001811475  1091121844806                       0.689   
123  2001811305  1091121846136                       0.750   

     Weight slab as per X (KG)  Total weight as per Courier Company (KG)  \
0                          1.5                                      1.30   
1                          1.0                                      1.00   
2                          2.5                                      2.50   
3                          1.0                                      1.00   
4                          0.5                                      0.15   
..                         ...                                       ...   
119                        0.5                                      0.73   
120                        1.0                                      0.50   
121                        1.0                                      0.50   
122                        1.0                                      0.50   
123                        1.0                                      0.50   

     Weight slab charged by Courier Company (KG) Delivery Zone as per X  \
0                                            1.5                      d   
1                                            1.0                      d   
2                                            2.5                      d   
3                                            1.0                      b   
4                                            0.5                      d   
..                                           ...                    ...   
119                                          1.0                      b   
120                                          0.5                      b   
121                                          0.5                      b   
122                                          0.5                      e   
123                                          0.5                      b   

    Delivery Zone charged by Courier Company  Expected Charge as per X (Rs.)  \
0                                          d                           135.0   
1                                          d                            90.2   
2                                          d                           224.6   
3                                          b                            61.3   
4                                          d                            45.4   
..                                       ...                             ...   
119                                        d                            33.0   
120                                        d                           100.7   
121                                        d                            61.3   
122                                        b                           112.1   
123                                        d                            61.3   

     Charges Billed by Courier Company (Rs.)  
0                                      135.0  
1                                       90.2  
2                                      224.6  
3                                       61.3  
4                                       45.4  
..                                       ...  
119                                     90.2  
120                                     86.7  
121                                     45.4  
122               

Obs- We can see the 10th column is added 

#### * To add the last column i.e "Difference Between Expected Charges and Billed Charges (Rs.)" we just have to perform subtraction

In [71]:
data1['Difference Between Expected Charges and Billed Charges (Rs.)']=round((data1['Expected Charge as per X (Rs.)']-data1['Charges Billed by Courier Company (Rs.)']),4)

In [72]:
data1

OrderID      AWBNumber  Total weight as per X (KG)  \
0    2001806232  1091117222124                       1.302   
1    2001806273  1091117222194                       0.615   
2    2001806408  1091117222931                       2.265   
3    2001806458  1091117223244                       0.700   
4    2001807012  1091117229345                       0.240   
..          ...            ...                         ...   
119  2001812941  1091118551656                       0.500   
120  2001809383  1091117614452                       0.607   
121  2001820978  1091120922803                       0.515   
122  2001811475  1091121844806                       0.689   
123  2001811305  1091121846136                       0.750   

     Weight slab as per X (KG)  Total weight as per Courier Company (KG)  \
0                          1.5                                      1.30   
1                          1.0                                      1.00   
2                          2.5                                      2.50   
3                          1.0                                      1.00   
4                          0.5                                      0.15   
..                         ...                                       ...   
119                        0.5                                      0.73   
120                        1.0                                      0.50   
121                        1.0                                      0.50   
122                        1.0                                      0.50   
123                        1.0                                      0.50   

     Weight slab charged by Courier Company (KG) Delivery Zone as per X  \
0                                            1.5                      d   
1                                            1.0                      d   
2                                            2.5                      d   
3                                            1.0                      b   
4                                            0.5                      d   
..                                           ...                    ...   
119                                          1.0                      b   
120                                          0.5                      b   
121                                          0.5                      b   
122                                          0.5                      e   
123                                          0.5                      b   

    Delivery Zone charged by Courier Company  Expected Charge as per X (Rs.)  \
0                                          d                           135.0   
1                                          d                            90.2   
2                                          d                           224.6   
3                                          b                            61.3   
4                                          d                            45.4   
..                                       ...                             ...   
119                                        d                            33.0   
120                                        d                           100.7   
121                                        d                            61.3   
122                                        b                           112.1   
123                                        d                            61.3   

     Charges Billed by Courier Company (Rs.)  \
0                                      135.0   
1                                       90.2   
2                                      224.6   
3                                       61.3   
4                                       45.4   
..                                       ...   
119                                     90.2   
120                                     86.7   
121                                     45.4   
122     

Obs- We have got our final column , i.e difference between expected charges and billed charges. The values that are -ve is the extra amount levied by their Delivery partners per Order

In [89]:
# Creating our first ouput file i.e Data1 in excel
data1.to_excel('Order Level Calculation.xlsx')

## Creating 2nd Output file: Summary File

In [74]:
Summary=pd.DataFrame({' ' : ["Total orders where X has been correctly charged", "Total Orders where X has been overcharged", "Total Orders where X has been undercharged"],'Count':'','Amount (Rs.)':''})

In [75]:
Summary

Count Amount (Rs.)
0  Total orders where X has been correctly charged                   
1        Total Orders where X has been overcharged                   
2       Total Orders where X has been undercharged

In [76]:
Summary.index=Summary[' ']

In [77]:
del Summary[" "]

In [78]:
Summary

Count Amount (Rs.)
                                                                  
Total orders where X has been correctly charged                   
Total Orders where X has been overcharged                         
Total Orders where X has been undercharged

In [79]:
# Filling the values 0ne by one for the created cross-tab
# 1. Count[0]
Summary.loc[:]['Count'][0]=data1["Difference Between Expected Charges and Billed Charges (Rs.)"][data1["Difference Between Expected Charges and Billed Charges (Rs.)"]==0].value_counts()[0]

In [80]:
Summary

Count Amount (Rs.)
                                                                  
Total orders where X has been correctly charged    21             
Total Orders where X has been overcharged                         
Total Orders where X has been undercharged

In [81]:
#2. Amount[0]
Summary.loc[:]['Amount (Rs.)'][0]=data1["Charges Billed by Courier Company (Rs.)"][data1["Difference Between Expected Charges and Billed Charges (Rs.)"]==0].sum()

In [82]:
Summary

Count Amount (Rs.)
                                                                  
Total orders where X has been correctly charged    21       1765.6
Total Orders where X has been overcharged                         
Total Orders where X has been undercharged

In [83]:
# 3. Count[1]
Summary.loc[:]['Count'][1]=data1["Difference Between Expected Charges and Billed Charges (Rs.)"][data1["Difference Between Expected Charges and Billed Charges (Rs.)"]<0].value_counts().sum()
Summary

Count Amount (Rs.)
                                                                  
Total orders where X has been correctly charged    21       1765.6
Total Orders where X has been overcharged          91             
Total Orders where X has been undercharged

In [84]:
# 4. Amount[1]
Summary.loc[:]['Amount (Rs.)'][1]=abs(data1["Difference Between Expected Charges and Billed Charges (Rs.)"][data1["Difference Between Expected Charges and Billed Charges (Rs.)"]<0].sum())
Summary

Count Amount (Rs.)
                                                                  
Total orders where X has been correctly charged    21       1765.6
Total Orders where X has been overcharged          91       5013.8
Total Orders where X has been undercharged

In [85]:
# 5.Count[2]
Summary.loc[:]['Count'][2]=data1["Difference Between Expected Charges and Billed Charges (Rs.)"][data1["Difference Between Expected Charges and Billed Charges (Rs.)"]>0].value_counts().sum()
Summary

Count Amount (Rs.)
                                                                  
Total orders where X has been correctly charged    21       1765.6
Total Orders where X has been overcharged          91       5013.8
Total Orders where X has been undercharged         12

In [86]:
# Amount[2]
Summary.loc[:]['Amount (Rs.)'][2]=abs(data1["Difference Between Expected Charges and Billed Charges (Rs.)"][data1["Difference Between Expected Charges and Billed Charges (Rs.)"]>0].sum())
Summary

Count Amount (Rs.)
                                                                  
Total orders where X has been correctly charged    21       1765.6
Total Orders where X has been overcharged          91       5013.8
Total Orders where X has been undercharged         12        495.5

#### * Conclusion- 
We have got the complete summary of extra amount levied by Company X's Delivery partners.
Out of the total 124 total orders delivered by the courier company, 21 orders have been fairly charged, 12 orders have been undercharged, but a good number of orders i.e 91 have been overcharged

In [87]:
# Converting the summary table into excel file

In [88]:
Summary.to_excel('Summary.xlsx')

##### Thank You for this Assignment 